### PR TITLE
Move macros to `detail/__config`

### LINKED ISSUE #143: Create a `detail/__config` file for common configuration definitions
Note to self: We should make a `detail/__config` file for this kind of thing.

_Originally posted by @jrhemstad in https://github.com/NVIDIA/cuCollections/pull/98#discussion_r833275984_

### PR DESCRIPTION
**Pull request by sleeepyjack**
Closes #143

### CODE REVIEW

File: include/cuco/detail/__config [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#if defined(CUDART_VERSION) && (CUDART_VERSION >= 11000) && defined(__CUDA_ARCH__) && \
+  (__CUDA_ARCH__ >= 700)
```
Comment (PointKernel) [inline include/cuco/detail/__config:20]: This can use the `and` logic of `CUCO_HAS_CG_MEMCPY_ASYNC` and `CUCO_HAS_LARGE_TYPE_SUPPORT`

Comment (PointKernel) [reply include/cuco/detail/__config:20]: Actually, I'm not sure if we want this macro depending on `CUCO_HAS_CG_MEMCPY_ASYNC` and `CUCO_HAS_LARGE_TYPE_SUPPORT`.

Comment (sleeepyjack) [reply include/cuco/detail/__config:20]: Yes, would be possible but I don't see the semantic relation between these flags.

Comment (PointKernel) [reply include/cuco/detail/__config:20]: Agreed.

**Review by PointKernel — APPROVED** [review]

File: include/cuco/detail/__config [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+#define CUCO_HAS_CUDA_BARRIER
+#endif
+
+#if defined(CUDART_VERSION) && (CUDART_VERSION >= 11100)
+#define CUCO_HAS_CG_MEMCPY_ASYNC
+#endif
+
+#if defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 700)
+#define CUCO_HAS_LARGE_TYPE_SUPPORT
```
Comment (jrhemstad) [inline include/cuco/detail/__config]: I think a more appropriate name would be `CUCO_HAS_VOLTA_THREAD_MODEL` or `CUCO_HAS_INDEPENDENT_THREADS`.

The "large type support" is a consequence of needing to use locking, which needs the independent thread model.

Comment (jrhemstad) [reply include/cuco/detail/__config]: This reminds me that in order to be friendly to nvc++, we should probably be switching to use `NV_IF_TARGET`. That'll be a different issue though.

**Comment by jrhemstad** [comment]
@PointKernel I thought you had made a similar `CUCO_HAS_INDEPENDENT_THREADS` macro before? Or am I just imagining that?

**Comment by PointKernel** [comment]
Here we go https://github.com/NVIDIA/cuCollections/pull/107#discussion_r740452805.

I remember I did something similar and then find this thanks to your reminder:
https://github.com/PointKernel/cuCollections/blob/3a49fc7143366e33334ce25846fb9854fbdc99f3/tests/CMakeLists.txt#L67-L72.

Should we move the CMake function to a higher level CMake file thus the macro can be used by both headers and test files (i.e. both device and host code)?

**Comment by jrhemstad** [comment]
> Should we move the CMake function to a higher level CMake file thus the macro can be used by both headers and test files (i.e. both device and host code)?

I don't think such a thing is possible.

How would we detect when someone is using the `cuco` CMake target in a library that specifies the `sm_60` arch?

**Comment by sleeepyjack** [comment]
Good question * *scratches head* *. It would be nice though to have a mechanism like this. For the Bloom filter for instance I need something like `CUCO_HAS_L2_RESIDENCY_CONTROL` detectable from the host in order to throw a sensible error in case someone tries to enable this feature for <sm_80.

**Comment by jrhemstad** [comment]
@robertmaynard says this may actually be possible. He's going to explore a few things.

**Comment by robertmaynard** [comment]
By using the `TARGET_PROPERTY` generator expression we can get CMake to evaluate what CUDA architectures the consuming target is being built for. Supporting the `all` , `all-major`, and `native` special keyword values make this harder but still possible.

The end result to do this detection at CMake time looks like:

```cmake
add_library(cuco INTERFACE)

# produce a list that has elements that match 60 / 60-real
# all or all-major. If the list has any elements it will
# convert to a boolean value of true
set(property "$<TARGET_PROPERTY:CUDA_ARCHITECTURES>")
set(filter_regex "$<FILTER:${property},INCLUDE,(60|all)>")
set(has_sm60 "$<BOOL:${filter_regex}>")

# support magic keyword native. If CUDA_ARCHITECTURES is set to native
# and CMake has detected a sm_6* hardware we enable the flag
set(is_native_mode "$<STREQUAL:native,${property}>")
set(native_regex "$<FILTER:${CMAKE_CUDA_ARCHITECTURES_NATIVE},INCLUDE,60>")
set(has_native_sm60 "$<AND:$<BOOL:${native_regex}>,${is_native_mode}>")


target_compile_definitions(cuco INTERFACE
  "$<${has_sm60}:HAVE_SM60>"
  "$<${has_native_sm60}:HAVE_SM60>"
  )
```

**Comment by jrhemstad** [comment]
:exploding_head:

**Comment by sleeepyjack** [comment]
Wizardry. Let me try that. I would say we let CMake expose `CUCO_HAS_SM_60_OR_HIGHER` (quite a mouthful; suggestions are appreciated) and then use this to define `CUCO_HAS_INDEPENDENT_THREADS` in the `__config` header.

**Comment by sleeepyjack** [comment]
One problem arises if we define the macro from the outside in order to use it as a switch in host code as in https://github.com/NVIDIA/cuCollections/blob/3a49fc7143366e33334ce25846fb9854fbdc99f3/tests/CMakeLists.txt#L67-L72

If we have a list of multiple target architectures, e.g., `sm_60,sm_80`, then the macro would be defined for all architectures. Since the current default set of target architectures also includes `sm_60`, the `STATIC_MAP_TEST` will always use the `CUCO_NO_INDEPENDENT_THREADS` code path.

Moreover, if we use this definition here
https://github.com/NVIDIA/cuCollections/blob/3a49fc7143366e33334ce25846fb9854fbdc99f3/include/cuco/static_map.cuh#L159-L164
we would disable large type support, even on Volta+ GPUs.

I would suggest we don't move the `CUCO_NO_INDEPENDENT_THREADS` definition to a higher level CMakeFile and keep it as a WAR inside the test folder.

**Comment by PointKernel** [comment]
> Since the current default set of target architectures also includes sm_60

Ah, I misunderstood what you were referring to about the "default set of archs" during the offline discussion so sorry for the misleading information. cuCo uses `NATIVE` build by default:https://github.com/NVIDIA/cuCollections/blob/a63731fdf464909069c98cf20d96d361bec7f3cc/CMakeLists.txt#L52 thus it won't compile for all architectures defined in `rapids-cmake`

> we would disable large type support, even on Volta+ GPUs.

If the local system has only one `sm_70` GPU, the default build with `CMAKE_CUDA_ARCHITECTURES=NATIVE` won't generate a sm_60 compatible binary. If a system has `sm_60` as well as `sm_70`, it should be the users' responsibility to explicitly specify the desired target architecture e.g. `CMAKE_CUDA_ARCHITECTURES=70` to get the best performance.

**Comment by jrhemstad** [comment]
The problem is that a library using cuco (like libcudf) compiles with `sm_60,sm_70,...` to support multiple GPUs in the shipped binary.

**Comment by PointKernel** [comment]
> The problem is that a library using cuco (like libcudf) compiles with `sm_60,sm_70,...` to support multiple GPUs in the shipped binary.

Based on offline discussion with @sleeepyjack , we will leave the test `CUCO_NO_INDEPENDENT_THREADS` macro as it is to avoid the issue mentioned here.

**Review by PointKernel — APPROVED** [review]

**Review by jrhemstad — APPROVED** [review]

**Pull request merged**